src/int/specialized_div_rem/mod.rs

```diff
@@ -0,0 +1,314 @@
+// TODO: when `unsafe_block_in_unsafe_fn` is stabilized, remove this
+#![allow(unused_unsafe)]
+
+//! This `specialized_div_rem` module is from version 0.4.0 of the `specialized-div-rem`
+//! crate. The build system cannot handle including it as a regular dependency, so we have
+//! pasted `asymmetric.rs`, `binary_long.rs`, `delegate.rs`, `norm_shift.rs`, and
+//! `trifecta.rs` here without modification. Note that `for` loops with ranges are not
+//! used in this module, since unoptimized compilation may generate references to `memcpy`.
```

Comment:
Originally I wanted to make `compiler-builtins` depend on `specialized_div_rem`, but I was told the build system cannot handle `compiler-builtins` depending on anything currently. I have never messed with git submodules before, and it appears that git does not download submodules by default which would bring its own complications. I resorted to pasting in code. Whenever we want to change the algorithmic code, the least hacky way to do it is to modify the original code on my `specialized_div_rem` master branch, and then paste it in without needing to apply further modifications. I have set things up such that most modifications only need to touch `mod.rs` to tweak configuration flags and not touch the algorithmic code.

Reply:
My problem with the vendor-by-copy approach is that these codebases will inevitably diverge over time and just "copy paste form upstream" will eventually become unfeasible approach to keeping both in-sync.

Reply:
The implementation can be entirely owned by `compiler-builtins`, and any future contributor can change the code, but I also want the algorithmic macros in my `specialized-div-rem` crate to be able to be identical to that in `compiler-builtins`. Maybe I should just let go and say that if any future contributor wants to break this dependency they can. I don't know if I will ever touch my `specialized-div-rem` crate again. Should I change the wording to something like: "This code is originally from `specialized-div-rem` 0.4.0".

Reply:
> Maybe I should just let go and say that if any future contributor wants to break this dependency they can. I don't know if I will ever touch my specialized-div-rem crate again.

If the very good implementation is made available in `compiler-builtins`, the external crate kind of loses its purpose, except maybe as a place to experiment on further improvements, so that would make sense to me.

If we end up allowing the code in `compiler-builtins` to eventually diverge from what’s in the crate, then I would probably take this approach:

1. Reference the crate where the code comes from in e.g. `mod.rs` once, as you suggested;
2. Add yourself to the list of authors in `Cargo.toml` if you want your contribution to be named/acknowledged;
3. Land the code as is without caring much about the minor stylistic nits I made.

After the PR lands we can then clean up the code to match the compiler-builtins conventions better, without needing to worry about problems that the divergence would pose, as the codebases would effectively be entirely independent from the landing point.